doc/source/development/window.rst

```diff
@@ -0,0 +1,242 @@
+Generalized Window Operations
+=============================
+
+Rationale
+---------
+The current window implementation in :file:`window.pyx` is difficult to
+customize and couples the creation of window bounds, loop over the values to
+aggregate, the logic for updating an aggregation. The aggregation algorithms
+are also duplicated in :file:`groupby.pyx`, but unifying grouping aggregation
+algorithms and windowing aggregation algorithms is out of scope for this
+proposal
+
+Requirements
+------------
+#. Pass all existing user-facing API tests
+#. User-defined aggregations (UDAs) performance on par with builtin aggregations
+#. Ideally no performance regressions
+#. Performant support for custom offsets (e.g., custom business
+   days)
+#. New features are strict additions to the API
+
+Soft Requirements
+-----------------
+#. Lower the maintenance burden of the existing implementation by using numba
+   if possible
+#. Reduce code duplication
+
+Implementation
+--------------
+Here we describe details of the implementation that we think satisfy the
+requirements.
+
+Ideally, the implementation is able to use ``numba`` to implement the
+performance sensitive parts of the proposal.
```

Comment:
We should probably justify why we want to use Numba. I think the reasons are:
* Easy debugging
* OOB support for types needed in Pandas.
* Easy implementation of compiled UDAs.